relay-server/src/actors/envelopes.rs

```diff
@@ -1197,8 +1298,20 @@ impl EnvelopeProcessor {
         let filter_settings = &state.project_state.config.filter_settings;
 
         metric!(timer(RelayTimers::EventProcessingFiltering), {
-            relay_filter::should_filter(event, client_ip, filter_settings)
-                .map_err(ProcessingError::EventFiltered)
+            relay_filter::should_filter(event, client_ip, filter_settings).map_err(|err| {
+                if let Some(ref outcome_producer) = self.outcome_producer {
+                    EnvelopeContext::new(
```

Comment:
Shouldn't the envelope context be passed in here? I'd suggest moving all this functionality to the `ProcessEnvelopeState`. Here, you only need to give the impulse of "drop envelope", and the state should already contain all information to then emit the outcome.

The same goes for sampling. Ideally, we keep the EnvelopeContext and pass it around, rather than reconstructing it at every outcome site.